presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,245 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.sql.planner.LogicalPlanner;
+import com.facebook.presto.sql.planner.Partitioning;
+import com.facebook.presto.sql.planner.Plan;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.assertions.BasePlanTest;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.testing.TestngUtils.toDataProvider;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.Files.exists;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.testng.Assert.assertEquals;
+
+public abstract class BaseCostBasedPlanTest
+        extends BasePlanTest
+{
+    private final Path mavenModulePath;
+
+    public BaseCostBasedPlanTest(String mavenModule, LocalQueryRunnerSupplier supplier)
+    {
+        super(supplier);
+        this.mavenModulePath = Paths.get(requireNonNull(mavenModule, "mavenModulePath is null"));
+    }
+
+    protected abstract Stream<Query> getQueries();
+
+    @DataProvider
+    public Object[][] getQueriesDataProvider()
+    {
+        return getQueries()
+                .collect(toDataProvider());
+    }
+
+    @Test(dataProvider = "getQueriesDataProvider")
+    public void test(Query query)
+    {
+        String actual = joinOrderString(resolve(query.query))
+                .stream()
+                .collect(joining("\n"));
+        assertEquals(actual, read(resolve(getExpectedJoinOrderingFile(query.id))));
+    }
+
+    private Path resolve(Path path)
+    {
+        if (exists(path)) {
+            return path;
+        }
+        Path fromParent = Paths.get("..").resolve(path);
```

Comment:
Why do we need this? Why don't specify an exact patch where the expected results are laying?

Reply:
We need this because when test is run, its current working directory is the test's module. However, when update (eg `TestTpchCostBasedPlan.UpdateTestFiles#main`) is run from Intellij, it's cwd is the project's top folder. I will reinstate this method.

Reply:
Why should we care about working directory? Why don't just resolve the resources via the classloader? We store the resources in the classpath? Right? 

Also, if you ever want to run this test (or generator) packaged in the Jar, it will fail. Always resolve resources via the class loader.